### ISSUE TITLE
Ocusync 4 cannot be resolved

### ISSUE THREAD

**Issue by wanhongwu1998** [proto17/dji_droneid#50]
Hello, thank you very much for your sharing. Our research has found that this burst does not exist in LightBrige technology, so it is impossible to parse such a drone. In addition, although the burst can be found in Ocusync 4 technology and the constellation image can be demodulated successfully, the signal of the drone cannot be solved, as it seems to be encrypted. I wonder if you have encountered the same problem in the future.

**Comment by proto17** [comment]
I have not done any work with newer drones. If you have a collect that you're willing to share I would be happy to have a look (keeping in mind your physical location could be present in the collect!), but the only drone I have is a Mini 2. If the constellation comes out clean and you get data back then I would suggest looking at that data with a hex editor to see if there are any strings present. IIRC the serial number of the drone shows up in the demodulated output. If you see that serial number in the output then you can be pretty sure that the link isn't encrypted. It could be that the CRC changed, or that the format of the demodulated output changed. But it's definitely still possible that the link has been encrypted. There's been lots of calls for that to happen. 

It could also be that you have a really noisy collect that's failing to properly demodulate. The `process_file.m` script doesn't output the frame hex if the CRC fails. You can change this by commenting out [1]. Doing so will output the hex for all frames even if they are not valid according to the CRC. 

Good luck!

[1] https://github.com/proto17/dji_droneid/blob/main/cpp/remove_turbo.cc#L156

**Comment by wanhongwu1998** [comment]
Thank you for your reply. I'm sorry that it is not convenient to export the data I collected to you! To be sure, CRC was successful, so I'm more confident that the format of the demodulated output changed or was encrypted.

**Comment by jianjianliu0918** [comment]
> Thank you for your reply. I'm sorry that it is not convenient to export the data I collected to you! To be sure, CRC was successful, so I'm more confident that the format of the demodulated output changed or was encrypted.

DJI officially said it was encrypted, but recently a company claimed it had successfully decrypted it.

**Comment by wanhongwu1998** [comment]
> > Thank you for your reply. I'm sorry that it is not convenient to export the data I collected to you! To be sure, CRC was successful, so I'm more confident that the format of the demodulated output changed or was encrypted.
> 
> DJI officially said it was encrypted, but recently a company claimed it had successfully decrypted it.

I doubt they can parse encrypted data. If encrypted data can still be parsed, then DJI's engineers are useless, and of course it doesn't rule out that they have superior abilities or have some special deal with DJI.

**Comment by ozymandiasleo** [comment]
> Hello, thank you very much for your sharing. Our research has found that this burst does not exist in LightBrige technology, so it is impossible to parse such a drone. In addition, although the burst can be found in Ocusync 4 technology and the constellation image can be demodulated successfully, the signal of the drone cannot be solved, as it seems to be encrypted. I wonder if you have encountered the same problem in the future.

Did you revolve ocusync 3.0 ?

**Comment by alphafox02** [comment]
Even if the signal can’t be decrypted, there’s still value in knowing it exists. If a drone signal is within range of your receiver, and you deploy additional receivers positioned at strategic distances, maybe you could triangulate the signal’s location based on the strength detected by each sensor.

This assumes the signal can still be identified, even if it’s encrypted. I believe the AntSDR e200 firmware works similarly—it detects encrypted signals and flags them as such, without decoding the contents.

**Comment by 1264763882** [comment]
您好 我想问一下ocusync4是通过时间加密的吗 我的qpsk簇已经解析的比较好了 但解码部分误码严重

**Comment by jianjianliu0918** [comment]
> 您好 我想问一下ocusync4是通过时间加密的吗 我的qpsk簇已经解析的比较好了 但解码部分误码严重

国标SM2加密的，不用想着破解了

**Comment by 1264763882** [comment]
我观察dji协议无人机数据解扰送进去的数据去除zc序列后固定从4148位置送入1412个bit码元 解码位置是否有变化

**Comment by RenDJunyu** [comment]
> 您好 我想问一下ocusync4是通过时间加密的吗 我的qpsk簇已经解析的比较好了 但解码部分误码严重

DJI的droneid应该是可以解调的，但是图传和跳频应该是加密的

**Comment by Umairnxr** [comment]
Hello, I am working on the Ocu4 Drone ID decoding. is there anyone who can help me out. I have a mini 4 pro.

**Comment by alexandrethomas-flyingeye** [comment]
Hi everyone, does anyone have the O4 decryption key from DJI? I'm looking to purchase it.

**Comment by Sanafan** [comment]
anyone able to sell decryption key or server?

**Comment by wanhongwu1998** [comment]
I can now obtain some data that has passed the CRC test but is encrypted. I know the plaintext corresponding to several frames of the encryption. However, I don't know how it was transformed and I'm not sure if you are still working on it? How can I get in touch with you?

> @proto17 #

**Comment by luyii-code-1** [comment]
A small update on O4: I have been working on a reproducible O2/O4 DroneID receiver and packet-analysis project, and the encryption structure is now considerably clearer than when this issue was originally opened.

Repository:
https://github.com/luyii-code-1/dji-ocusync-droneid-research

Cross-validation thread:
https://github.com/luyii-code-1/dji-ocusync-droneid-research/issues/1

For DJI Mini 5 Pro captures I can recover O4 AA/87 logical packets with both CRC24A and DJI CRC16 passing.

The currently verified chain is:

```text
AA
  -> SM2-compatible C1 || C3 || C2
  -> 16-byte note

87
  -> AES-128-CTR
  -> SN / UUID / drone / pilot / home telemetry
```

The AA key-material region has a 64 + 32 + 16 byte layout matching raw-point SM2 C1||C3||C2, and C1 validates on the SM2 curve across multiple sessions.

The 16-byte `note` from the paired AA is the AES-128 key for 87. With the correct note, sequential 87 packets decrypt into coherent telemetry using:

`IV = nonce8 || 0x00*8`

AA and same-session 87 packets also share the same 4-byte hashcode.

So the remaining hard part is now specifically `AA -> note`, likely requiring the corresponding SM2 private key or an equivalent decryptor, rather than understanding the 87 ciphertext format itself.

I would be very interested in independent O4 captures from other models/firmware for cross-validation.

**Comment by Hellikandra** [comment]
> A small update on O4: I have been working on a reproducible O2/O4 DroneID receiver and packet-analysis project, and the encryption structure is now considerably clearer than when this issue was originally opened.
> 
> Repository: https://github.com/luyii-code-1/dji-ocusync-droneid-research
> 
> Cross-validation thread: [luyii-code-1/dji-ocusync-droneid-research#1](https://github.com/luyii-code-1/dji-ocusync-droneid-research/issues/1)
> 
> For DJI Mini 5 Pro captures I can recover O4 AA/87 logical packets with both CRC24A and DJI CRC16 passing.
> 
> The currently verified chain is:
> 
> ```
> AA
>   -> SM2-compatible C1 || C3 || C2
>   -> 16-byte note
> 
> 87
>   -> AES-128-CTR
>   -> SN / UUID / drone / pilot / home telemetry
> ```
> 
> The AA key-material region has a 64 + 32 + 16 byte layout matching raw-point SM2 C1||C3||C2, and C1 validates on the SM2 curve across multiple sessions.
> 
> The 16-byte `note` from the paired AA is the AES-128 key for 87. With the correct note, sequential 87 packets decrypt into coherent telemetry using:
> 
> `IV = nonce8 || 0x00*8`
> 
> AA and same-session 87 packets also share the same 4-byte hashcode.
> 
> So the remaining hard part is now specifically `AA -> note`, likely requiring the corresponding SM2 private key or an equivalent decryptor, rather than understanding the 87 ciphertext format itself.
> 
> I would be very interested in independent O4 captures from other models/firmware for cross-validation.

Independent cross‑validation from a separate O3/O4 analysis project — I can confirm your AA/87 model byte‑for‑byte from a different codebase:

CRYP 0xAA = SM2 point C1‖C3‖C2 = 64+32+16 = 112 bytes, 4‑byte hashcode
INFP 0x87 = AES‑128‑CTR, IV = nonce8 ‖ 0x00×8, same 4‑byte hashcode as its CRYP
CRYP↔INFP linked by that hashcode; keyed‑INFP decrypt reproduces coherent SN/lat/lon/alt/pilot/home

Our detector/parser uses the identical constants and passes both CRC‑24A and the DJI CRC‑16, so this is now reproduced by two independent implementations. Happy to send independent O4 captures from other models (Avata 2 / Goggles 3) for cross‑firmware validation — I can export a JSON packet dump (CRYP split into C1/C3/C2, INFP nonce+ciphertext, hashcode, CRC status) that should drop straight into your pipeline.

One correction that I think saves effort, re: "the material to decrypt the session key is embedded in both the drone and the RC" — for the SM2 layer that isn't the case, and it's worth being precise because it changes where to look:

CRYP is the session key encrypted with DJI's SM2 public key. Unwrapping it needs the SM2 private key, which by design is not on the drone or the RC — the drone only holds the public key (to encrypt), and DroneID is a broadcast for remote‑ID, so the RC never decrypts CRYP at all (it already has telemetry over the control link). So firmware strings/RE will surface DJI's public key and the AES‑CTR/hashcode code, but not a CRYP‑decrypting private key — it isn't there to find.
The only on‑device route is the other end: the aircraft generated the 16‑byte session key itself, so it holds the plaintext key transiently in RAM before wrapping it. Recovering a session's telemetry therefore means lifting that 16‑byte key from the aircraft's own memory (a root shell / hardware access on the drone SoC), not decrypting CRYP. That's the realistic (if heavy) path; the AA → note step stays walled by SM2 (ECDLP, ~2¹²⁸) without DJI's private key.
Also worth separating: the DroneID broadcast session key (SM2‑wrapped, per‑flight) is distinct from the control‑link binding session key (the OcuSync data channel). Conflating them sends you looking for the wrong key in the wrong place.

Net: the 87 format is fully understood (agreed), and AA → note is a hard asymmetric wall — reachable only via DJI's private key or the aircraft's RAM, not via RC/firmware key material. Glad to contribute the cross‑model captures whenever useful.

**Comment by Sanafan** [comment]
I have many drones... 

M400
M30T
4x M4TD
M4E
Mavic 4 Pro


How to contribute?

**Comment by luyii-code-1** [comment]
The plane isn't necessary, mainly the SDR equipment



---Original---
From: ***@***.***&gt;
Date: Fri, Sep 25, 2026 01:09 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [proto17/dji_droneid] Ocusync 4 cannot be resolved (Issue #50)


Sanafan left a comment (proto17/dji_droneid#50)
 
I have many drones...
 
M400
 M30T
 4x M4TD
 M4E
 Mavic 4 Pro
 
How to contribute?

**Comment by Sanafan** [comment]
> The plane isn't necessary, mainly the SDR equipment
> 
> 
> 
> 
> 
> 
> 
> ---Original---
> 
> From: ***@***.***&gt;
> 
> Date: Fri, Sep 25, 2026 01:09 AM
> 
> To: ***@***.***&gt;;
> 
> Cc: ***@***.******@***.***&gt;;
> 
> Subject: Re: [proto17/dji_droneid] Ocusync 4 cannot be resolved (Issue #50)
> 
> 
> 
> 
> 
> Sanafan left a comment (proto17/dji_droneid#50)
> 
>  
> 
> I have many drones...
> 
>  
> 
> M400
> 
>  M30T
> 
>  4x M4TD
> 
>  M4E
> 
>  Mavic 4 Pro
> 
>  
> 
> How to contribute?
> 
>  
> 
> —
> 
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 
> You are receiving this because you commented.Message ID: ***@***.***&gt;

I also have an ant sdr e200

**Comment by luyii-code-1** [comment]
Maybe we could try to get some "AA/80" packets. Even though we've pretty much confirmed the encryption method now, cross-validation can never hurt



---Original---
From: ***@***.***&gt;
Date: Fri, Sep 25, 2026 01:17 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [proto17/dji_droneid] Ocusync 4 cannot be resolved (Issue #50)


Sanafan left a comment (proto17/dji_droneid#50)
  
The plane isn't necessary, mainly the SDR equipment
 
---Original---
 
From: @.***&gt;
 
Date: Fri, Sep 25, 2026 01:09 AM
 
To: @.***&gt;;
 
Cc: @.@.&gt;;
 
Subject: Re: [proto17/dji_droneid] Ocusync 4 cannot be resolved (Issue #50)
 
Sanafan left a comment (#50)
 
I have many drones...
 
M400
 
M30T
 
4x M4TD
 
M4E
 
Mavic 4 Pro
 
How to contribute?
 
—
 
Reply to this email directly, view it on GitHub, or unsubscribe.
 
You are receiving this because you commented.Message ID: @.***&gt;
  
I also have an ant sdr e200

**Comment by Hellikandra** [comment]
Here is maybe an idea : 

**Goal:** build a multi‑model / multi‑firmware dataset of DJI **OcuSync 4** DroneID
packets so the **AA/87 structure** can be independently confirmed across the whole O4
line (Mini 4, Air 3/3S, Mavic 4, Avata 2, Matrice 4/30/300‑class, …).

This is **receive‑only, no flight required, no keys required.** You are only recording
and parsing the *structure* of the public remote‑ID broadcast — the telemetry payload
stays encrypted (that's expected; see "What this does / doesn't do").

Model confirmed so far (Mini 5 Pro, independent decoders agree):

```
AA (CRYP) -> SM2 point C1||C3||C2 = 64+32+16 = 112 bytes, + 4‑byte hashcode
87 (INFP) -> AES‑128‑CTR, IV = nonce8 || 0x00*8, same 4‑byte hashcode as its AA
```

---

## TL;DR

```bash
# EITHER just share your raw capture:
#   grab.s16.iq  +  the metadata block below      (we/others demodulate it)
#
# OR, if you can demodulate to bytes, run the ONE self-contained file (no dependencies):
python3 o4_droneid_parse.py demod.bin --json o4_pkts.json --model "DJI Avata 2" --firmware "V01.00.0100"
# then share o4_pkts.json
```

---

## Requirements

- **An SDR** that reaches 2.4 and 5.8 GHz: ADALM‑Pluto, **AntSDR E200**, USRP B2xx, etc.
- **Python 3** — nothing else for the parser.
- **`o4_droneid_parse.py`** — a **single, self-contained** file (stdlib only, no numpy, no other
  project files). This is the only file you need to produce the standardized dump; it is included
  alongside this README.

> The maintainers use a larger `sdr_droneid_o4.py` (adds an IQ *detector* and a keyed-decrypt test,
> which need `numpy`/`cryptography`). Contributors don't need it — `o4_droneid_parse.py` produces the
> identical JSON from any demodulated byte stream.

---

## Two ways to contribute (pick your level)

### Level 1 — raw IQ grab (easiest; anyone with an SDR)

Just record the drone's DroneID emission and share the file + metadata. We/others demodulate it.

1. Power the drone (a GPS lock helps populate the telemetry, but a bench capture is fine).
2. Record **~5 s** of IQ at **15.36 MS/s** on the DroneID band (2.4 **and** 5.8 GHz if you can):
   - Pluto/AntSDR example rate: `15.36e6`. Centre on the operating channel; DroneID hops, so a
     wide grab (AntSDR's larger bandwidth is ideal) catches more.
   - Save as interleaved **int16 I,Q** (`.s16.iq`) — the format the tools read.
3. Share `grab.s16.iq` (or a trimmed clip) **plus the metadata block below**. That's it —
   no tool needed at your end for Level 1.

### Level 2 — parsed packet dump (if you have a working O4 receiver)

If you already demodulate O4 DroneID to bytes (e.g. with the luyii receiver, or your own),
run the single self‑contained parser to produce the standardized JSON:

```bash
python3 o4_droneid_parse.py demod.bin --json o4_pkts.json --model "..." --firmware "..."
```

`demod.bin` = the demodulated DroneID byte stream (the 0x55‑framed logical packets).
The parser scans it, validates **CRC‑8 (seed 0x77)** and **CRC‑16 (seed 0x3692)** on each
frame, identifies **CRYP (0xAA)** / **INFP (0x87)**, links them by the 4‑byte hashcode, and
splits the CRYP SM2 point into **C1(64)‖C3(32)‖C2(16)** — no dependencies.

---

## The JSON format you share

```json
{
  "source": "grab.bin",
  "tool": "sdr_droneid_o4",
  "constants": { "CRYP":"0xaa","INFP":"0x87","sm2_point_len":112,
                 "layout":"C1(64)||C3(32)||C2(16)","hashcode_len":4,
                 "infp_iv":"nonce8||0x00*8","cipher":"AES-128-CTR" },
  "n_packets": N, "n_cryp": N, "n_infp": N, "n_linked": N,
  "packets": [
    { "type":170,"kind":"CRYP (SM2-wrapped session key)","hashcode":"aabbccdd",
      "sm2_point":"...112 bytes hex...","sm2_C1":"...64...","sm2_C3":"...32...","sm2_C2":"...16...",
      "encrypted":true },
    { "type":135,"kind":"INFP (AES-128-CTR telemetry)","hashcode":"aabbccdd",
      "nonce":"...8 bytes hex...","ciphertext":"...hex...","linked_cryp":true,"encrypted":true }
  ]
}
```

## Metadata to include with every submission

```
model:            e.g. DJI Avata 2 / Mavic 4 Pro / M30T / M4TD
firmware version: e.g. V01.00.0100  (aircraft + RC if known)
band / centre:    2.4 GHz 2444.x  or  5.8 GHz 5xxx
sdr / rate:       AntSDR E200 @ 15.36 MS/s
capture notes:    GPS locked? indoor/outdoor? distance? duration?
```

---

## What this does / doesn't do

- **Does:** confirm the AA/87 packet structure, the C1‖C3‖C2 layout, the hashcode linkage,
  and the CRC framing across models/firmware. That is genuinely useful and needs no keys.
- **Doesn't:** decrypt the telemetry. The INFP content (serial / position / operator) is
  **AES‑128‑CTR** under a key that is **SM2‑wrapped with DJI's public key** — unwrapping it needs
  DJI's SM2 **private** key, which is not on any device. The dump you share is **ciphertext +
  framing only**, so it exposes **no plaintext PII**.

---

*Tool: `o4_droneid_parse.py` — one self‑contained file, stdlib only. Its JSON is receiver‑agnostic:
it works on any correctly demodulated O4 DroneID byte stream, whatever produced it. (Maintainers'
`sdr_droneid_o4.py` adds an IQ detector + keyed‑decrypt test and produces the identical dump.)*

```python
#!/usr/bin/env python3
"""o4_droneid_parse.py — STANDALONE O4 DroneID packet parser (no dependencies).

Single-file, pure-Python (stdlib only) version of the CRYP/INFP parser, meant to be
shared with contributors so they need NOTHING else — no numpy, no other project files.

It scans a demodulated DroneID byte stream for 0x55-framed logical packets whose
CRC-8 (seed 0x77) and CRC-16 (seed 0x3692) both pass, identifies CRYP(0xAA)/INFP(0x87),
links them by 4-byte hashcode, splits the CRYP SM2 point into C1||C3||C2 (64/32/16),
and writes a standardized JSON dump for cross-model validation.

    python3 o4_droneid_parse.py demod.bin --json o4_pkts.json

`demod.bin` = the demodulated DroneID packet bytes (from any working O4 receiver).
Produces ciphertext + framing only — no plaintext PII (the telemetry stays encrypted).
"""
import sys, json, struct, argparse

PKT_CRYP = 0xAA      # SM2-wrapped 16-byte AES session key
PKT_INFP = 0x87      # AES-128-CTR telemetry
SM2_POINT_LEN = 112  # C1(64) || C3(32) || C2(16)
HASHCODE_LEN = 4
INFP_NONCE_LEN = 8   # AES-CTR IV = nonce8 || 0x00*8

def crc8(data, c=0x77):
    for b in data:
        c ^= b
        for _ in range(8):
            c = (c >> 1) ^ 0x8C if (c & 1) else (c >> 1)
    return c & 0xFF

def crc16(data, c=0x3692):
    for b in data:
        c ^= b
        for _ in range(8):
            c = (c >> 1) ^ 0x8408 if (c & 1) else (c >> 1)
    return c & 0xFFFF

def parse_o4_packets(data):
    data = bytes(data); out = []; i = 0
    while i < len(data) - 6:
        if data[i] != 0x55:
            i += 1; continue
        total = data[i + 1] | (data[i + 2] << 8)
        if total < 9 or i + total > len(data):
            i += 1; continue
        frame = data[i:i + total]
        if crc8(frame[:3]) != frame[3]:
            i += 1; continue
        if crc16(frame[:-2]) != (frame[-2] | (frame[-1] << 8)):
            i += 1; continue
        ptype = frame[4]
        hashcode = frame[5:5 + HASHCODE_LEN]
        payload = frame[5 + HASHCODE_LEN:-2]
        rec = dict(type=ptype, hashcode=hashcode.hex(), length=total, payload_len=len(payload))
        if ptype == PKT_CRYP:
            pt = payload[:SM2_POINT_LEN]
            rec.update(kind="CRYP (SM2-wrapped session key)", encrypted=True,
                       sm2_point=pt.hex(), sm2_C1=pt[:64].hex(), sm2_C3=pt[64:96].hex(), sm2_C2=pt[96:112].hex(),
                       note="session key SM2-encrypted; needs DJI's SM2 PRIVATE key (not on device)")
        elif ptype == PKT_INFP:
            rec.update(kind="INFP (AES-128-CTR telemetry)", encrypted=True,
                       nonce=payload[:INFP_NONCE_LEN].hex(), ciphertext=payload[INFP_NONCE_LEN:].hex(),
                       note="telemetry AES-128-CTR, IV=nonce8||0x00*8; needs the 16-byte session key")
        else:
            rec["kind"] = "other"
        out.append(rec); i += total
    cryps = {r["hashcode"] for r in out if r["type"] == PKT_CRYP}
    for r in out:
        if r["type"] == PKT_INFP and r["hashcode"] in cryps:
            r["linked_cryp"] = True
    return out

def main():
    ap = argparse.ArgumentParser(description="Standalone O4 DroneID CRYP/INFP parser (no deps)")
    ap.add_argument("bin", help="demodulated DroneID packet byte file")
    ap.add_argument("--json", metavar="OUT", help="write the standardized packet dump")
    ap.add_argument("--model", default="", help="drone model (goes in the JSON metadata)")
    ap.add_argument("--firmware", default="", help="firmware version (metadata)")
    a = ap.parse_args()
    pkts = parse_o4_packets(open(a.bin, "rb").read())
    for p in pkts:
        print(p)
    n_cryp = sum(1 for p in pkts if p.get("type") == PKT_CRYP)
    n_infp = sum(1 for p in pkts if p.get("type") == PKT_INFP)
    n_link = sum(1 for p in pkts if p.get("linked_cryp"))
    print(f"\n{len(pkts)} CRC-valid packets: {n_cryp} CRYP, {n_infp} INFP, {n_link} linked")
    if not pkts:
        print("(no CRC-valid O4 packets — check the demod / byte alignment)")
    if a.json:
        meta = dict(source=a.bin, tool="o4_droneid_parse", model=a.model, firmware=a.firmware,
                    constants=dict(CRYP=hex(PKT_CRYP), INFP=hex(PKT_INFP), sm2_point_len=SM2_POINT_LEN,
                                   layout="C1(64)||C3(32)||C2(16)", hashcode_len=HASHCODE_LEN,
                                   infp_iv="nonce8||0x00*8", cipher="AES-128-CTR",
                                   crc8_seed="0x77", crc16_seed="0x3692"),
                    n_packets=len(pkts), n_cryp=n_cryp, n_infp=n_infp, n_linked=n_link, packets=pkts)
        json.dump(meta, open(a.json, "w", encoding="utf-8"), indent=2)
        print(f"[json] wrote {a.json}")
    return 0 if pkts else 1

if __name__ == "__main__":
    sys.exit(main())
```
